src/structstore_shared.hpp

```diff
@@ -115,8 +247,23 @@ class StructStoreShared {
     }
 
     ~StructStoreShared() {
-        if (--shm_ptr->usage_count == 0) {
-            // unlink shared memory
+
+        size_t usage_count = 0;
+
+        if (shm_ptr != nullptr) {
+
+            if (reinit) {
+                shm_ptr->ok.store(false);
+            }
+
+            shm_ptr->usage_count -= 1;
+            usage_count = shm_ptr->usage_count;
```

Comment:
I'm not sure why this is necessary. In any case, these two lines should be done in a single statement as it was done before to make proper use of the atomic integer:
```suggestion
            usage_count = shm_ptr->usage_count -= 1;
```